flow/ActorCollection.actor.cpp

```diff
@@ -21,26 +21,71 @@
 #include "flow/ActorCollection.h"
 #include "flow/IndexedSet.h"
 #include "flow/UnitTest.h"
+#include <boost/intrusive/list.hpp>
 #include "flow/actorcompiler.h" // This must be the last #include.
 
+struct Runner : public boost::intrusive::list_base_hook<>, FastAllocated<Runner>, NonCopyable {
+	Future<Void> handler;
+};
+
+// An intrusive list of Runners, which are FastAllocated.  Each runner holds a handler future
+typedef boost::intrusive::list<Runner, boost::intrusive::constant_time_size<false>> RunnerList;
+
+// The runners list in the ActorCollection must be destroyed when the actor is destructed rather
+// than before returning or throwing
+struct RunnerListDestroyer : NonCopyable {
+	RunnerListDestroyer(RunnerList* list) : list(list) {}
+
+	~RunnerListDestroyer() {
+		list->clear_and_dispose([](Runner* r) { delete r; });
+	}
+
+	RunnerList* list;
+};
+
+ACTOR Future<Void> runnerHandler(PromiseStream<RunnerList::iterator> output,
+                                 PromiseStream<Error> errors,
+                                 Future<Void> task,
+                                 RunnerList::iterator runner) {
+	try {
+		wait(task);
+		output.send(runner);
+	} catch (Error& e) {
+		if (e.code() == error_code_actor_cancelled)
+			throw;
+		errors.send(e);
+	}
+	return Void();
+}
+
 ACTOR Future<Void> actorCollection(FutureStream<Future<Void>> addActor,
                                    int* pCount,
                                    double* lastChangeTime,
                                    double* idleTime,
                                    double* allTime,
                                    bool returnWhenEmptied) {
-	state int64_t nextTag = 0;
-	state Map<int64_t, Future<Void>> tag_streamHelper;
-	state PromiseStream<int64_t> complete;
+	state RunnerList runners;
+	state RunnerListDestroyer runnersDestroyer(&runners);
+	state PromiseStream<RunnerList::iterator> complete;
 	state PromiseStream<Error> errors;
 	state int count = 0;
 	if (!pCount)
 		pCount = &count;
 
 	loop choose {
 		when(Future<Void> f = waitNext(addActor)) {
-			int64_t t = nextTag++;
-			tag_streamHelper[t] = streamHelper(complete, errors, tag(f, t));
+			// Insert new Runner at the end of the instrusive list and get an iterator to it
+			auto i = runners.insert(runners.end(), *new Runner());
+
+			// Start the handler for completions or errors from f, sending runner to complete stream
+			Future<Void> handler = runnerHandler(complete, errors, f, i);
+
+			// If the list reverse begin is still i, which is what it would be after the insertion above,
+			// then f did not complete or error so store the handler future in the list item to keep it alive.
+			if (std::next(runners.rbegin()).base() == i) {
```

Comment:
It actually turns out that the reason I thought `i` _could_ be invalidated here is due to confusing crash behavior when I was debugging a version of this where the list destruction was happening explicitly before the `return` or `throw` in the main loop.  After starting the handler, even if it completes immediately because `f` was already done/errored, since we are in one of the `when` blocks the callbacks for the other blocks have already been removed so any send to the `complete` or `errors` streams are just queued to be processed on the next loop iteration, so i->handler can just be unconditionally set.  Thanks @sfc-gh-anoyes for digging into this.

Reply:
To complete the story, creating the list auto destructor on scope closure was the fix for the actual bug I was hitting.